### DISCUSSION TITLE
Typescript globals are not working [Solved]

### DISCUSSION THREAD

**Discussion by Alino** [vitest-dev/vitest#1573] [General]
I have followed the instructions to add vitest/globals into tsconfig so that typescript should recognise globals. But it does not.
expect.toEqual is missing.

tsconfig.json

```json
{
    "compilerOptions": {
      "types": ["node", "vitest/globals"],
      "lib": ["ES2015"],
      "esModuleInterop": true
    }
}
```

vitest.config.ts
```typescript
/// <reference types="vitest" />
import { defineConfig } from 'vite'

export default defineConfig({
  test: {
    globals: true,
    exclude: [
        "folder-dev",
    ]
  },
})
```

test file:
<img width="923" alt="image" src="https://user-images.githubusercontent.com/763194/176564031-e2315d71-2cb8-434a-9432-4c940c20ea9c.png">

Any idea what's wrong?

**Comment by sheremet-va** [comment]
You have other global expect type contradicting with ours

**Reply by Alino** [reply]
thanks I was thinking that too, so I have removed them (jest and jest globals) from tsconfig. Do I have to remove jest from package.json as well? I wanted to keep jest and vitest so that I could use them both during a migration phase.

**Reply by sheremet-va** [reply]
`toEqual` exists on both. I think you have mocha expect globally somewhere. Do you use cypress?

**Reply by Alino** [reply]
I don't use cypress. These were my dev dependencies including _jest_, _mocha_, _ts-jest_, _@types/jest_ and _@jest/globals_

```json
"devDependencies": {
    **"@jest/globals": "^27.5.1",**
    **"@types/jest": "^27.4.1",**
    **"@types/mocha": "^9.0.0",**
    "@types/node": "^16.11.26",
    "@types/rosie": "^0.0.40",
    "@types/supertest": "^2.0.11",
    **"jest": "^28.1.1",**
    **"mocha": "^9.1.3",**
    "rosie": "^2.1.0",
    "supertest": "^6.1.6",
    **"ts-jest": "^28.0.0-next.3",**
    "ts-node": "^10.7.0",
    "typescript": "^4.7.4",
    "vitest": "^0.16.0"
  }
```

so I have removed now my node_modules and removed these so now I have

```json
  "devDependencies": {
    "@types/node": "^16.11.26",
    "@types/rosie": "^0.0.40",
    "@types/supertest": "^2.0.11",
    "rosie": "^2.1.0",
    "supertest": "^6.1.6",
    "ts-node": "^10.7.0",
    "typescript": "^4.7.4",
    "vitest": "^0.16.0"
  }
```

run ```yarn install```

and I still got the issue. Also restarted vscode.

EDIT: Btw, running ```npx vitest``` executes the tests and all are passing fine. It's just the vscode's typescript somehow thinking wrong?

**Comment by Alino** [comment]
EDIT: I was wrong about thinking that it works on my other machine, because it was just initialising there and not yet underlining the issue... So the same issue is on both instances of vscode. Even in a dev docker container.

EDIT2: So I found the root cause. In the root of my repo I had index.d.ts with:
```typescript
declare const expect: Chai.ExpectStatic
```
I totally didn't know about this there lol.

**Comment by daribock** [comment]
I have the same problem but didn't declare any const or anything. I searched through every file but I found nothing.
This is the error I'am getting:
<img width="638" alt="image" src="https://user-images.githubusercontent.com/57658125/190401287-4cd6792b-b1d8-4239-8ffd-0f3d8fbaa3a2.png">


`tsconfig.vitest.json`
```Code
{
    "extends": "./tsconfig.json",
    "compilerOptions": {
        "types": ["vitest/globals"]
    },
    "include": ["src/**/*.test.ts"]
}
```

`vite.config.ts`
```Code
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import eslintPlugin from 'vite-plugin-eslint';
import i18nHotReload from './plugins/i18n-reload';

export default defineConfig((configEnv) => ({
    plugins: [react(), eslintPlugin(), i18nHotReload()],
    server: {
        port: 3000,
    },
    test: {
        globals: true,
        environment: 'jsdom',
        include: ['src/**/*.test.{ts,tsx}'],
        coverage: {
            reporter: ['text'],
        },
    },
}));
```

Do you have any idea what I'am missing? @sheremet-va  😕

**Reply by daribock** [reply]
Forget this. In my `.tsconfig.json` was this config:
```code
"exclude": ["src/__mocks__", "src/**/*.test.ts"],
```

But now I have the same issue as @Alino but I use Cypress in my project but until know its only installed and the types are not configured yet. I removed it but it still didnt work.

**Comment by yashdodani** [comment]
When I did 'vitest run', it ran all the tests without any errors, but vscode still showed errors for 'test', 'expect' or 'describe'.
this error is due to eslint
installing eslint-plugin-vitest-globals and setting .eslintrc.cjs
`module.exports = {
    root: true,
    env: { browser: true, es2020: true, 'vitest-globals/env': true },
    extends: [
        'eslint:recommended',
        'plugin:react/recommended',
        'plugin:react/jsx-runtime',
        'plugin:react-hooks/recommended',
        'plugin:vitest-globals/recommended',
    ],`
this solved the problem for me.

**Reply by ekim49** [reply]
This worked for me. Thanks!

**Reply by 13RTK** [reply]
If you create a TypeScript + React app with vite, attach the types in tsconfig.app.json instead of tsconfig.json

tsconfig.json:
```json
{
  "files": [],
  "references": [
    { "path": "./tsconfig.app.json" },
    { "path": "./tsconfig.node.json" }
  ]
}
```

tsconfig.app.json:
```json
{
  "compilerOptions": {
    "tsBuildInfoFile": "./node_modules/.tmp/tsconfig.app.tsbuildinfo",
    "target": "ES2022",
    "useDefineForClassFields": true,
    "lib": ["ES2022", "DOM", "DOM.Iterable"],
    "module": "ESNext",
    "types": ["vite/client", "vitest/globals"],
    "skipLibCheck": true,

    "paths": {
      "@/*": ["./src/*"]
    },

    /* Bundler mode */
    "moduleResolution": "bundler",
    "allowImportingTsExtensions": true,
    "verbatimModuleSyntax": true,
    "moduleDetection": "force",
    "noEmit": true,
    "jsx": "react-jsx",

    /* Linting */
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "erasableSyntaxOnly": true,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedSideEffectImports": true
  },
  "include": ["src"]
}
```

vite.config.ts:
```typescript
/// <reference types="vitest/config" />

import react from '@vitejs/plugin-react';
import { defineConfig } from 'vitest/config';
import { fileURLToPath, URL } from 'node:url';
import { playwright } from '@vitest/browser-playwright';

export default defineConfig({
  plugins: [react()],
  resolve: {
    alias: {
      '@': fileURLToPath(new URL('./src', import.meta.url)),
    },
  },
  test: {
    browser: {
      enabled: true,
      provider: playwright(),
      instances: [{ browser: 'chromium' }],
    },
    globals: true,
  },
});
```
